Rewrite the grammar

amaanq opened the pull request:
- Reduces state count (tbh, I had it greatly reduced to around ~7-8k, but a few grievances grew it back to nearly what it was before, ~12k, further optimizations can come once I dig more later)
- Fully implements C# 12 and some of 13, including primary constructors, ref readonly parameters, and aliasing *any* type.
- Fixes parsing bugs related to interpolated strings when more than one `$` is used by delegating this to the external scanner for statefulness by leveraging a stack of the current interpolation's info (yes, nested interpolations are handled nicely, no, it was not fun at times to implement)
- Adds new bindings for other languages that upstream supports, tidied up manifests
- CI is more robust using new upstream workflows
- Improved queries
- Updated tests, and removing duplicates/excessive tests. There's no need for one test per line of code/tree sitter rule, it just adds unnecessary higher-level nodes when several of these tests can be combined into one test to test them all together
- Rewrote grammar to be more in-tune with the tree-sitter style, including getting rid of useless precedences, rules, hidden or not, inlining where appropriate, supertyping relevant rules, unhiding certain rules/supertypes, etc.
- Improved preproc calls by having content be nested inside these nodes, but this does break the case where it's in the middle of an if statement for example, similar to C. I think the tradeoff is worth it, only ~80 files out of 8k+ files fail because of this, and we get much nicer parse trees in the common case, which, e.g., improves code folding. Another added benefit is we can use preproc rules in spots where only that rule itself is valid, and not everything. e.g. a preproc containing expressions is valid inside expressions, but this won't be done at the top level.
- Fuzz the scanner for obvious reasons

I will update the tokens for publishing the new version later today (needs pypi).

- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/189
- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/283 (todo - quotes)
- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/310
- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/319
- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/326
- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/331
- Closes https://github.com/tree-sitter/tree-sitter-c-sharp/issues/332

tamasvajk reviewed (commented):
Thank you for this fix/improvement/contribution. This is going to be awesome. 

I find this PR rather difficult to review. I think the first commit could also contain the test output changes. That would make it easier to see if anything has been broken.

I think there are a couple of bugs that are (re)introduced with this refactoring.
- some of the contextual keyword don’t seem to work any longer, for example parameter `p` in `void m(scoped p) { }` doesn’t have a type. Probably `scoped` became a modifier.
- `var o = (Int32)(1);` became an invocation with a parenthesised identifier, instead of a cast. https://github.com/tree-sitter/tree-sitter-c-sharp/issues/175
- `var greetings = new (string head, string tail)[3];` is not an array creation expression in the equals clause. https://github.com/tree-sitter/tree-sitter-c-sharp/issues/166

There are a lot of changes that I don't understand. Can you explain what's the benefit of changing these:
- some field names have been removed or changed.
- `null_literal` is gone, so `throw null` is now a `throw_expression` without any child.  
- `equals_value_clause` has been removed from the grammar,
- `type_parameter_constraints_clause` has been reworked, previously we had a list of clauses with target and constraint, now we have a flat list of constraints, in the form of `identifier, type_parameter_constraint+, identifier, type_parameter_constraint+`. Isn't this more difficult to handle in for consumers of the output tree?
- `with_initializer_expression` is gone, which means `with_expression` contains a flat list of children. This way the tree seems to be more difficult to process.

amaanq commented:
Wow, those two dynamic precedence issues caught me off guard. I noticed some places where prec.dynamic was used unnecessarily, but I didn't think those two were. That was really interesting and after a little digging, it makes sense. This is probably one of few grammars that is very sensitive to dynamic precedence changes. Thanks for the context, I've added them back

Others:

- A lot of the fields were unnecessary, however, if you think some of the ones i removed were necessary, let me know and I can add them back
- I added this back, it's generally preferable to not assign a node to a single literal since it just "absorbs" that literal and you can't directly query for the string literal then, but for consistency with all the other literals it makes sense
- I don't see the point in this, unless it resolved ambiguities/reduced state count but I don't think it did
- I remember this *initially* reduced the state count a lot when I was starting my rewrite, but now I noticed it doesn't, so yeah I can just have it as it was before
- Added back, a field called `initializer` could work though if we just want to tag/distinguish the identifier though

hvitved commented on grammar.js:
```diff
@@ -32,237 +43,304 @@ const stringEncoding = /(u|U)8/;
-      $.record_declaration,
-      $.record_struct_declaration
+        seq(
+          repeat(alias($.statement, $.global_statement)),
```
This change means that `node-types.json` does not provide any information about what `global_statement` can be. Can we revert it back to `repeat($.global_statement)` with `global_statement: $ => $.statement`?

amaanq replied:
I could, but what's the benefit? Curious

hvitved replied:
We have [a generic Treesitter-based CodeQL extractor framework](https://github.com/github/codeql/blob/main/shared/tree-sitter-extractor/src/generator/mod.rs), which, given a Treesitter grammar (`node-types.json`), produces a matching CodeQL database schema. For example, given `node-types.json` [for Ruby](https://github.com/tree-sitter/tree-sitter-ruby/blob/master/src/node-types.json), we produce [a matching DB schema](https://github.com/github/codeql/blob/main/ruby/ql/lib/ruby.dbscheme#L105). When I point it at the `node-types.json` file in this PR, it currently fails because it doesn't know the type of `global_statement`, and hence it is not able to generate a type union that represent all the possible nodes under a `compilation_unit`. In comparison, the [Ruby `program` rule](https://github.com/tree-sitter/tree-sitter-ruby/blob/master/grammar.js#L108) gives rise to [this](https://github.com/github/codeql/blob/main/ruby/ql/lib/ruby.dbscheme#L997) union type.

hvitved replied:
I guess one way we can work around this is basically to say then that a `compilation_unit` can contain _any_ type of AST node as a child.

amaanq replied:
Gotcha, thanks. I'm okay with adding the rule back, though it would be a smidge better without it (one less reduction)

hvitved replied:
In that case I think it would be nice to add it back in, thanks.

damieng commented on Cargo.toml:
```diff
@@ -1,36 +1,27 @@
-    "Patrick Thomson <patrickt@github.com>",
-    "Noelle Caldwell <noelle.caldwell@microsoft.com>",
-    "Douglas Creager <dcreager@dcreager.net>",
+  "Max Brunsfeld <maxbrunsfeld@gmail.com>",
```
Why are we wiping some authors out here?

amaanq replied:
I consider the Cargo.toml authors list to be those that are relevant to the **Rust** crate and core grammar, and can be reached out to for any questions, comments, or concerns. Some of those authors either a) have no direct relation to the rust crate, b) are potentially unreachable/inactive.

amaanq commented:
Well I wasn't aware I had to discuss making large improvements beforehand, nor have I ever done so in any other upstream grammar I maintain.

The moving of tests to a different dir is *necessary*, top-level corpus dirs are unsupported upstream now and must be in `test/corpus`.

The exclusion list is either a parse error where it genuinely was an error (beforehand as well) or it has funky usage of preproc ifs, e.g. in the middle of an if/else statement. I have (imo) improved them such that their contents are children of the `preproc_if` node, much like how C does it, however, this requires that the contents inside be somewhat correctly formed (e.g. a regular statement/expression, wherever the preproc_if is *applicable*). I think this is a reasonable tradeoff for much more navigable trees.

amaanq commented:
@damieng @tamasvajk the diff for tests should be much more readable now

maxbrunsfeld commented on test/corpus/literals.txt:
```diff
@@ -128,71 +128,63 @@ var x = $"{{";
         (variable_declarator
           name: (identifier)
-          (equals_value_clause
-            (integer_literal))))))
```
👍 to getting rid of `equals_value_clause`.

maxbrunsfeld commented on test/corpus/classes.txt:
```diff
@@ -83,119 +89,114 @@ file class A {}
     (modifier)
     name: (identifier)
-    body: (declaration_list))
+    (declaration_list))
```
IMO the `body` field was fine, but I don't know if there was some drawback to it. Curious if you were seeing significant code size increases due to that field.

damieng replied:
There's a few of these removed in places that @tamasvajk added - are they used by GH Semantic?

tamasvajk replied:
I think the fields are mostly used by GH CodeQL. @hvitved knows best how much we need them.

hvitved replied:
Having field names in general certainly makes the life easier for Treesitter based QL extractors. Field names are used to generate methods in CodeQL, e.g. [this field](https://github.com/tree-sitter/tree-sitter-ruby/blob/9d86f3761bb30e8dcc81e754b81d3ce91848477e/grammar.js#L196) in the Ruby grammar gives rise to [this QL predicate](https://github.com/github/codeql/blob/main/ruby/ql/lib/codeql/ruby/ast/internal/TreeSitter.qll#L1443), which is nicer to use than something like `getChild(0)`.

maxbrunsfeld commented on queries/highlights.scm:
```diff
@@ -179,77 +195,19 @@
-(switch_expression (identifier) @variable)
 
-;; Lock statement
-(lock_statement (identifier) @variable)
```
👍 to removing all of these specific patterns that match `@variable`

maxbrunsfeld commented:
Hope you're able to reduce the state count. In general, a great way to do that is avoid long sequences with many variations.

In other words, turn this:

```js
seq(
  optional('foo'),
  optional('bar'),
  optional('baz'),
  optional('quux'),
  // ...
)
```

into this:

```js
seq(
  repeat(choice('foo', 'bar', 'baz', 'quux')),
  // ...
)
```

It seems like there may be opportunities to do some of that in this grammar - there may be unnecessarily-specific sequences that could be modeled more generically.

damieng commented on test/corpus/attributes.txt:
```diff
@@ -8,15 +8,15 @@ Global attributes
     (attribute
       name: (identifier)
-      (attribute_argument_list))))
+      (attribute_arguments))))
```
These changes are going to be breaking and they're going to take us away from the Roslyn grammar definitions.

e.g. attribute_argument_list being wiped out and attribute_argument becoming attribute_arguments.

Originally I aligned with them because it made it easier to write the grammar but it's also been really helpful as the language evolves to see which rules they changed for the new features and how ours changes too. It's also provided a great way to know we're parsing the code correctly as we can match the definitive parser.

If moving away from that is desirable because it makes the grammar a lot smaller or enables some new tree-sitter functionality it might be worth the trade-off but if not I'd prefer to retain the alignment and compatibility.

damieng commented on test/corpus/contextual-keywords.txt:
```diff
@@ -188,9 +184,9 @@ class scoped { }
-          (parameter_modifier)
-          (parameter_modifier)
-          type: (predefined_type)
+          type: (scoped_type
```
This looks very off? They've gone from parameter modifiers to type modifiers?

damieng commented on test/corpus/expressions.txt:
```diff
@@ -19,63 +19,54 @@ a = a!;
     (expression_statement
       (assignment_expression
         left: (identifier)
-        (assignment_operator)
```
This one I'm fine with - it aligns better with Roslyn.

damieng commented on test/corpus/expressions.txt:
```diff
@@ -405,35 +366,29 @@ var x = new A
-          (identifier))
-        (assignment_operator)
-        right: (integer_literal)))))
+      (prefix_unary_expression
```
Subtle change here with the assignment now being a prefix_unary_expression rather than the left term.

amaanq replied:
good catch - that original splitting of _pointer_indirection_expression was very costly due to the conflict - but if we only allow lvalue_expression's there then it's much better

damieng commented on test/corpus/expressions.txt:
```diff
@@ -1167,63 +1088,53 @@ void m() {
-                  (block
-                    (return_statement
-                      (identifier))))))))
+              (anonymous_method_expression
```
Why is it we're losing "static" as a (modifier) here but we're keeping static as a modifier on the lambda equivalent on line 1262?

amaanq replied:
Gotcha, but what about static/unsafe in using_directive? I also (personally) think it's ok to hide modifiers, and just expose the string literals in the tree instead, but I don't have much of a preference.

damieng commented on test/corpus/enums.txt:
```diff
@@ -9,7 +9,7 @@ enum A: byte { One, Two = 2, Three = 0x03 }
   (enum_declaration
     name: (identifier)
-    bases: (base_list
+    (base_clause
```
base_list in Roslyn - not sure why we'd change it without a reason.

damieng commented on test/corpus/interfaces.txt:
```diff
@@ -129,34 +128,31 @@ public interface IGetNext<T> where T : IGetNext<T>
-                  (identifier)))
-              (variable_declarator
-                name: (identifier))))
+          (event_declaration
```
Again this has gone from a definitive Roslyn-defined set to what appears to be an arbitrary choice/level of abstraction.

Here's the section from syntax analyzer:

![image](https://github.com/tree-sitter/tree-sitter-c-sharp/assets/118951/124cde95-ba1b-4902-8181-a06e613d3b67)

amaanq replied:
no *this* one doesn't improve it all that much, to me I just thought that `event_declaration` satisfies this enough. this is innocuous enough to add back, but im curious why do we try to strictly follow the roslyn definition? I feel like that can cause some pain points in certain spots, but in this case it just adds a conflict that is sorta unnecessary, though state count only goes up a *smidge*

damieng replied:
Why would we not want to use the definitive name for the grammar parts rather than whoever adds a feature choose arbitrary names?

damieng commented on test/corpus/preprocessor.txt:
```diff
@@ -153,36 +149,36 @@ class Of1879 {
-            (preproc_string_literal))
-          (line_directive)
-          (line_directive))))))
+          (preproc_call
```
This feels like a definite step backwards in terms of usefulness.

amaanq replied:
do you mean the granular parsing of the line preprocessor directive? this is more how C handles it, but C# might be different because it has a fixed set of directives. I could add back the granularity, but I initially thought it's probably better to handle it like C

amaanq commented:
Down to 9700/3200 after reworking top level stuff + preprocs 😁

tamasvajk commented on test/corpus/literals.txt:
```diff
@@ -266,528 +250,518 @@ var x = $"{{";
-              (interpolation
-                (identifier))
-              (interpolated_string_text)))))))
+          (interpolated_string_expression
+            (interpolation_start)
+            (string_content)
+            (interpolation
+              (interpolation_open_brace)
+              (identifier)
+              (interpolation_close_brace))
+            (string_content))))))
```
Why do we include the start, quote and brace tokens in the tree?

amaanq replied:
they can be variable in size, i could hide them but this then becomes unqueryable, or i could alias them to a literal, e.g. alias open_brace to '{', but that wouldn't always be correct since it could be multiple

tamasvajk replied:
Makes sense. I'd keep it as it is. It just seemed unusual to see them in the tree.

tamasvajk reviewed (commented):
Thanks for rebasing the PR, it's a lot easier to check the changes in the test output. I've added some comments/questions.

amaanq commented on grammar.js:
```diff
@@ -281,148 +362,140 @@ module.exports = grammar({
       // 'scoped',  // `scoped` is either part of a scoped type or a scoped parameter. Both of which are handled outside of `modifier`.
       'sealed',
-      'static',
+      'static', // hmm?
```
I think I was just debugging something and left that as a note to self, oops

damieng commented on test/corpus/interfaces.txt:
```diff
@@ -170,7 +168,7 @@ public interface IGetNext<T> where T : IGetNext<T>
     body: (declaration_list
       (method_declaration
-        type: (predefined_type)
+        returns: (predefined_type)
```
@hvitved is this shift from type to returns going to cause upstream issues for semantic?

hvitved replied:
I don't work on Semantic, but rather CodeQL. We don't currently have an official tree-sitter based C# CodeQL extractor, so nothing breaks for us (as long as we can still correctly parse what we could earlier).

damieng commented on test/corpus/type-methods.txt:
```diff
@@ -247,21 +260,20 @@ ref struct S {
           (parameter
-            (parameter_modifier)
-            (parameter_modifier)
+            (modifier)
```
I'm okay with the parameter_modifier > modifier change

damieng commented on test/corpus/attributes.txt:
```diff
@@ -79,10 +79,9 @@ class Zzz {
-            (member_access_expression
-              expression: (identifier)
-              name: (identifier))))))
+          (member_access_expression
```
Maybe put back attribute_argument back between these two for consistency/non-breaking.

damieng commented on test/corpus/preprocessor.txt:
```diff
@@ -13,39 +13,32 @@ If, elif and else directives
-  (if_directive
-    (identifier))
-  (global_statement
+  (preproc_if
```
Let's see how we get on with these if > preproc changes - they're not part of Roslyn and probably less used by third party tooling relying on this library.

damieng commented on test/corpus/type-fields.txt:
```diff
@@ -125,52 +124,40 @@ struct A {
-                      (identifier))
-                    (function_pointer_unmanaged_calling_convention
-                      (identifier))))
+                (calling_convention
```
The function pointer calling convention has the old term in the grammar but it's definitely an less used construct so if this shared calling_convention results in an easier/shorter grammar I'm 👍🏻

damieng commented:
This is shaping up great! If we can get attribute_argument back in then I'm good on the Roslyn alignment/back compat from the naming side.

I think we just need to check/adjust the changes and removals of fields to make sure Semantic doesn't break and we should be good to go.

dcreager commented:
> I think we just need to check/adjust the changes and removals of fields to make sure Semantic doesn't break and we should be good to go.

Thanks for checking in, :+1: from our side!  We're pinned to the current release so we won't silently upgrade, and we're using the syntax highlighting queries directly from this repo, which I see are updated as part of this PR.  We have augmented the tagging queries with the ability to build up scoped names, which we'll have to update as part of bumping to the new version containing these changes.  But the changes to `tags.scm` in this PR look manageable, so I don't consider that a blocker.

amaanq commented:
thanks for the review/feedback @maxbrunsfeld @damieng @tamasvajk @hvitved!

hvitved commented:
> * Improved preproc calls by having content be nested inside these nodes, but this does break the case where it's in the middle of an if statement for example, similar to C. I think the tradeoff is worth it, only ~80 files out of 8k+ files fail because of this, and we get much nicer parse trees in the common case, which, e.g., improves code folding. Another added benefit is we can use preproc rules in spots where only that rule itself is valid, and not everything. e.g. a preproc containing expressions is valid inside expressions, but this won't be done at the top level.

Is there any way, using `extras` as before this PR, to still tolerate `#if`s that happen inside expressions or statements? For example, [this file ](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs) appears to have more parse errors now than before.
